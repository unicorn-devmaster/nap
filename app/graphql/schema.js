const { makeExecutableSchema } = require('graphql-tools')

const typeDefs = [`
type Query {
  getFoo(
    # Any String.
    bar: String!
  ): String
}

type Mutation {
  # Set something
  setFoo(
    # Any String.
    bar: String!
  ): String
}

schema {
  query: Query
  mutation: Mutation
}
`];

const resolvers = {
  Query: {
    getFoo: (root, { bar }, context) => {
      const displayName = (context.user && context.user.displayName)
      return `Hello ${bar} ${displayName ? `, Nice to see you ${displayName}!` : ', Please login'}`
    }
  },

  Mutation: {
    setFoo: (_, { bar }, context) => context.user ? `Saved! ${bar}` : `Aw!, you'll need to logged in first!`
  }
}

module.exports = makeExecutableSchema({
  typeDefs,
  resolvers,
})