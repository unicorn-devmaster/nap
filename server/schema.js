// Will use mongoose as default.
const mongoose = require('mongoose')
const { getSchema } = require('@risingstack/graffiti-mongoose')

mongoose.Promise = global.Promise
mongoose.connect(process.env.MONGO_URI || 'mongodb://mongo/graphql')

// Custom schema from models, hooks
const models = require('../models')()
const hooks = require('./hooks')
module.exports = getSchema(models, { hooks })
