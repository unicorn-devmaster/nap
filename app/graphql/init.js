const getSchema = (schemaURI, noCache) => {
  // Use no cache while development.
  noCache && delete require.cache[require.resolve(schemaURI)];

  // Serve
  return require(schemaURI)
}

const init = (app) => {
  // GraphQL's schema URI
  const GRAPHQL_SCHEMA = process.env.GRAPHQL_SCHEMA || './schema.js'
  
  // Custom config
  const IS_DEVELOPMENT = process.env.NODE_ENV === 'development'
  const GRAPHQL_PRETTY = (process.env.GRAPHQL_PRETTY === '1') || IS_DEVELOPMENT
  const GRAPHIQL_ENABLED = (process.env.GRAPHIQL_ENABLED === '1') || IS_DEVELOPMENT
  const GRAPHQL_SCHEMA_NOCACHE = (process.env.GRAPHQL_SCHEMA_NOCACHE === '1') || IS_DEVELOPMENT

  // Middleware
  const graphQLHTTP = require('express-graphql');
  const bodyParser = require('body-parser');

  // Apply
  app.use('/graphql',
    bodyParser.json(),
    graphQLHTTP((req, res) => ({
      graphiql: GRAPHIQL_ENABLED,
      pretty: GRAPHQL_PRETTY,
      schema: getSchema(GRAPHQL_SCHEMA, GRAPHQL_SCHEMA_NOCACHE)
    }))
  )
}

module.exports = init