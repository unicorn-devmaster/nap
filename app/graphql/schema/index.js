const mongoose = require('mongoose')
const { getSchema } = require('@risingstack/graffiti-mongoose')
const User = require(  './user')
const Pet = require(  './pet')
const config = require('../../config')

mongoose.Promise = global.Promise
mongoose.connect(config.mongo.url || 'mongodb://localhost/graphql')

module.exports = getSchema([Pet, User])