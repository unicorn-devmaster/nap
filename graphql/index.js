// - - - - - - Extras - - - - - -

require('./custom')

// - - - - - - GraphQL - - - - - -

// create new GQC from ComposeStorage
const { ComposeStorage } = require('graphql-compose')
const GQC = new ComposeStorage()

const { ErrorTC } = require('./ErrorSchema')
const { InstallationTC } = require('./InstallationSchema')
const { UserTC } = require('./UserSchema')
const { AuthenTC } = require('./AuthenSchema')

// ACL
const userAccess = (resolvers) => {
  Object.keys(resolvers).forEach((k) => {
    resolvers[k] = resolvers[k].wrapResolve(next => (rp) => {
      // rp = resolveParams = { source, args, context, info }
      if (!rp.context.nap.currentUser) {
        // throw new Error('[NOSTACK] Permission denied')
        rp.context.nap.errors.push({ code: 403, message: 'No session found' })
        return null
      }

      return next(rp)
    })
  })
  return resolvers
}

// create GraphQL Schema with all available resolvers for User Type
GQC.rootQuery().addFields(Object.assign(
  userAccess({
    // let add restriction for owner only
    user: UserTC.getResolver('user'),
  }), {
    authen: AuthenTC.getResolver('authen'),
    errors: ErrorTC.getResolver('error'),
  })
)

GQC.rootMutation().addFields(
  {
    signup: AuthenTC.getResolver('signup'),
    login: AuthenTC.getResolver('login'),
    logout: AuthenTC.getResolver('logout'),
    forget: AuthenTC.getResolver('forget'),
    loginWithFacebook: AuthenTC.getResolver('loginWithFacebook'),
    update_GCMSenderId: InstallationTC.getResolver('update_GCMSenderId'),
    update_deviceToken: InstallationTC.getResolver('update_deviceToken'),
    errors: ErrorTC.getResolver('error'),
  }
)

module.exports = GQC.buildSchema()
