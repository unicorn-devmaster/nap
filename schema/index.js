// Will use mongoose as default.
const mongoose = require('mongoose')
const { getSchema } = require('@risingstack/graffiti-mongoose')

mongoose.Promise = global.Promise
mongoose.connect(process.env.MONGO_URI || 'mongodb://mongo/graphql')

// Custom
const User = require('./user')
const Pet = require('./pet')

module.exports = getSchema([Pet, User])