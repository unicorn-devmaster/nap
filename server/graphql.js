const cors = require('cors')

const init = app => {
  // Custom config
  const graphiql = process.env.GRAPHIQL_ENABLED === '1' || process.env.NODE_ENV !== 'production'

  // Cross Origin
  app.use(cors())

  // Mongoose
  const graphqlHTTP = require('express-graphql')
  const { mainPage, addToMainPage } = require('./mainPage')
  const { getExampleNames, resolveExamplePath } = require('./composer')
  require('./mongooseConnection')

  const addExample = (example, uri) => {
    example.uri = `/${uri}`
    app.use(example.uri, graphqlHTTP(() => ({
      schema: example.schema,
      graphiql,
      formatError: (error) => ({
        message: error.message,
        stack: !error.message.match(/for security reason/i) ? error.stack.split('\n') : null,
      }),
    })))

    addToMainPage(example)
  }

  // scan `examples` directory and add
  // - graphql endpoint by uri /exampleDirName
  // - links and example queries to index page
  const exampleNames = getExampleNames()
  for (let name of exampleNames) {
    addExample(
      require(resolveExamplePath(name)).default,
      name
    )
  }

  app.get('/graphql-example', (req, res) => {
    res.send(mainPage())
  })
}

module.exports = init
