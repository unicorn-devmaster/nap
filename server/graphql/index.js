const { addResolverMiddleware, compose } = require('graphql-compose-recompose')
const config = require('../config')

const defaultBuildSchema = ({ GQC }) => {
  return GQC.buildSchema()
}

let buildGraphqlSchema = null

module.exports = {}
module.exports.getFile = (fileInput, context) => {
  if (!fileInput || !fileInput.mapKey) {
    return null
  }
  return context.files.find(file => {
    return file.originalname === fileInput.mapKey
  })
}
module.exports.extendModel = require('./models').extendModel
module.exports.setBuildGraphqlSchema = (builder) => (buildGraphqlSchema = builder)
module.exports.buildSchema = () => {
  let authenChannel
  async function loginMiddleware({ rp }, next) {
    const authen = await next()
    authenChannel.publish('login', { Authen_Id: authen._id.toString(), User_Id: authen.userId.toString(), Installation_Id: authen.installationId.toString() })
  }

  async function logoutMiddleware({ rp }, next) {
    const authen = await next()
    authenChannel.publish('logout', { Authen_Id: authen._id.toString(), User_Id: authen.userId.toString(), Installation_Id: authen.installationId.toString() })
  }

  const { ComposeStorage } = require('graphql-compose')
  const GQC = new ComposeStorage()
  const models = require('./models')()
  require('./composers')(models)
  const { onError } = require('../errors')

  const userAccess = (resolvers) => {
    Object.keys(resolvers).forEach((k) => {
      resolvers[k] = resolvers[k].wrapResolve(next => (rp) => {
        if (!rp.context.nap.session) {
          onError(rp.context)('No session found')
          return null
        }

        return next(rp)
      })
    })
    return resolvers
  }

  if (config.mubsub_enabled) {
    authenChannel = NAP.mubsub.client.channel('authen')

    models.AuthenTC = compose(
      addResolverMiddleware('login', loginMiddleware),
      addResolverMiddleware('loginWithFacebook', loginMiddleware),
      addResolverMiddleware('logout', logoutMiddleware)
    )(models.AuthenTC)
  }

  GQC.rootQuery().addFields(Object.assign(
    userAccess({
      user: models.UserTC.getResolver('user')
    }), {
      authen: models.AuthenTC.getResolver('authen'),
      errors: models.ErrorTC.getResolver('error')
    })
  )

  GQC.rootMutation().addFields(
    {
      signup: models.AuthenTC.getResolver('signup'),
      login: models.AuthenTC.getResolver('login'),
      logout: models.AuthenTC.getResolver('logout'),
      forget: models.AuthenTC.getResolver('forget'),
      loginWithFacebook: models.AuthenTC.getResolver('loginWithFacebook'),
      unlinkFacebook: models.UserTC.getResolver('unlinkFacebook'),
      linkFacebook: models.UserTC.getResolver('linkFacebook'),
      changeEmail: models.UserTC.getResolver('changeEmail'),
      update_GCMSenderId: models.InstallationTC.getResolver('update_GCMSenderId'),
      update_deviceToken: models.InstallationTC.getResolver('update_deviceToken'),
      errors: models.ErrorTC.getResolver('error')
    }
  )

  if (buildGraphqlSchema) {
    return buildGraphqlSchema({ GQC, models })
  }
  return defaultBuildSchema({ GQC })
}
