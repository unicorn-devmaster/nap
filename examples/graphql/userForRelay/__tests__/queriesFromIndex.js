import meta from '../index';
import schema from '../graphqlSchema';
import { graphql } from 'graphql';

function findQueryByTitle(str) {
  const queryConfig = meta.queries.find(o => o.title === str);
  if (queryConfig && queryConfig.query) {
    return queryConfig.query;
  }
  return 'query not found';
}

describe('userForRelay > queries', () => {
  const alwaysSameResultTitles = [
    'Relay node',
    'Relay Connection',
  ];
  alwaysSameResultTitles.forEach((title) => {
    it(title, async () => {
      const result = await graphql(meta.schema, findQueryByTitle(title));
      expect(result).toMatchSnapshot();
    });
  });

  {
    const title = 'Create user mutation';
    it(title, async () => {
      const result = await graphql(meta.schema, findQueryByTitle(title));
      expect(result.data.userCreate.record).toMatchSnapshot();
      expect(result.data.userCreate.clientMutationId).toMatchSnapshot();
    });
  }
});
