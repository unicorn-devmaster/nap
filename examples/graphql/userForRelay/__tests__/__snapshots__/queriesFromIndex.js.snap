exports[`userForRelay > queries Create user mutation 1`] = `
Object {
  "age": 24,
  "contacts": Object {
    "email": "mail@example.com",
    "phones": Array [
      "111-222-333-444",
      "444-555-666-777",
    ],
  },
  "gender": "ladyboy",
  "name": "My Name",
}
`;

exports[`userForRelay > queries Create user mutation 2`] = `"1"`;

exports[`userForRelay > queries Relay Connection 1`] = `
Object {
  "data": Object {
    "userConnection": Object {
      "edges": Array [
        Object {
          "cursor": "eyJfaWQiOiI1N2JiNDRkZDIxZDJiZWZiN2NhM2YwMDEifQ==",
          "node": Object {
            "_id": "57bb44dd21d2befb7ca3f001",
            "id": "VXNlclJlbGF5OjU3YmI0NGRkMjFkMmJlZmI3Y2EzZjAwMQ==",
            "name": "User 1",
          },
        },
        Object {
          "cursor": "eyJfaWQiOiI1N2JiNDRkZDIxZDJiZWZiN2NhM2YwMDIifQ==",
          "node": Object {
            "_id": "57bb44dd21d2befb7ca3f002",
            "id": "VXNlclJlbGF5OjU3YmI0NGRkMjFkMmJlZmI3Y2EzZjAwMg==",
            "name": "User 2",
          },
        },
        Object {
          "cursor": "eyJfaWQiOiI1N2JiNDRkZDIxZDJiZWZiN2NhM2YwMDMifQ==",
          "node": Object {
            "_id": "57bb44dd21d2befb7ca3f003",
            "id": "VXNlclJlbGF5OjU3YmI0NGRkMjFkMmJlZmI3Y2EzZjAwMw==",
            "name": "User 3",
          },
        },
      ],
    },
  },
}
`;

exports[`userForRelay > queries Relay node 1`] = `
Object {
  "data": Object {
    "node": Object {
      "_id": "57bb44dd21d2befb7ca3f010",
      "age": 21,
      "gender": "female",
      "id": "VXNlclJlbGF5OjU3YmI0NGRkMjFkMmJlZmI3Y2EzZjAxMA==",
      "name": "User 10",
    },
  },
}
`;
